src/main/java/com/google/firebase/database/connection/WebsocketConnection.java

```diff
@@ -333,17 +334,17 @@ public void run() {
 
     @Override
     public void onClose() {
-      final String logMessage = "closed";
-      executorService.execute(
-          new Runnable() {
-            @Override
-            public void run() {
-              if (logger.logsDebug()) {
-                logger.debug(logMessage);
-              }
-              onClosed();
-            }
-          });
+      if (logger.logsDebug()) {
+        logger.debug("closed");
+      }
+      if (!isClosed) {
+        executorService.execute(new Runnable() {
+          @Override
+          public void run() {
```

Comment:
I think you should probably duplicate the isClosed check here, since we're now on the correct thread (the one that mutates isClosed).

Reply:
`onClosed()` method has the check:

```
private void onClosed() {
    if (!isClosed) {
      if (logger.logsDebug()) {
        logger.debug("closing itself");
      }
      shutdown();
    }
    conn = null;
    if (keepAlive != null) {
      keepAlive.cancel(false);
      keepAlive = null;
    }
  }
```